R/clean_data_helpers.R

```diff
@@ -1,76 +1,147 @@
-#' Calculate the percentage of missing and other data type values in a vector
-#' containing different data types such as numeric, Date, character, and
-#' logical.
+#' Scan through all character columns of a data frame to determine the
+#' proportion of `missing`, `numeric`, `Date`, `character`, `logical`, values.
 #'
-#' @param x A vector of ones or a combination of various data types.
+#' @param data A data frame or linelist
+#'
+#' @returns A data frame with the same number of rows as the number of character
+#'    columns of the input data, and six (06) columns representing the field
+#'    names, the percentage of missing, numeric, date, character, and logical,
+#'    values in each column.
+#'
+#' @export
+#'
+#' @examples
+#' # scan through a data frame of characters
+#' scan_result <- scan_data(
+#'   data = readRDS(system.file("extdata", "messy_data.RDS",
+#'                              package = "cleanepi"))
+#' )
+#'
+#' # scan through a data frame with two character columns
+#' scan_result <- scan_data(
+#'   data = readRDS(system.file("extdata", "test_linelist.RDS",
+#'                              package = "cleanepi"))
+#'   )
+#'
+#' # scan through a data frame with no character columns
+#' data(iris)
+#' iris[["fct"]] <- as.factor(sample(c("gray", "orange"), nrow(iris),
+#'                            replace = TRUE))
+#' iris[["lgl"]]  <- sample(c(TRUE, FALSE), nrow(iris), replace = TRUE)
+#' iris[["date"]] <- as.Date(seq.Date(from = as.Date("2024-01-01"),
+#'                                    to = as.Date("2024-08-30"),
+#'                                    length.out = nrow(iris)))
+#' iris[["posit_ct"]] <- as.POSIXct(iris[["date"]])
+#' scan_result        <- scan_data(data = iris)
+#'
+scan_data <- function(data) {
+  # scan through all columns of the data and the identify character columns
+  types <- vapply(data, typeof, character(1L))
+  target_columns <- types[types == "character"]
+
+  # send an message if there is no character column found within the input data
+  if (length(target_columns) == 0L) {
+    message("No character column found in the provided data.")
+    return(invisible(NA))
+  }
+
+  # scan through the character columns
+  data <- data[, names(target_columns)]
+  scan_result <- vapply(seq_len(ncol(data)), function(col_index) {
+    scan_in_character(data[[col_index]])
+  }, numeric(5L))
+  scan_result        <- as.data.frame(t(scan_result))
```

Comment:
The reasoning behind this format is primarly due to the display in the report generated by `print_report()`.
For a dataset with many character columns, the user will have to scroll going to the right-side if the output of `scan_data()` is displayed as you shown above.
With cols -> rows, the first fifthteen columns are shown and the user can click on the available buttons to see the rest. This is simpler to my opinion - but I am happy to adjust if there is strong arguments for the other format.

Reply:
Thanks, rotating to avoid scrolling seems like a good reason! Can you make it explicit in the design vignette please?

Something like:

> We transpose the result relative to the input dataset (columns in the input are returned as row) to avoid horizontal scrolling in the case of datasets with a large number of character columns.